### PR TITLE
Implement AWS Signature Version 4 (yet another one)

### PR DESCRIPTION
**Pull request by sakazuki**
Hi,

I  rewrite S3Request.gs with AWS Signature Version 4.

Most code is ported from [aws-sdk-js/lib/signers/v4.js](https://github.com/aws/aws-sdk-js/blob/7cc9ae5b0d7b2935fa69dee945d5f3e6e638c660/lib/signers/v4.js)

Regards,

### CODE REVIEW

File: S3Request.gs [hunk 1]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+  if (this.isPresigned()) {
+    this.updateForPresigned(credentials, datetime);
+  } else {
+    this.addHeaders(credentials, datetime);
+  }
+  this.headers['Authorization'] = this.authorization(credentials, datetime)
+}
+
+S3Request.prototype.addHeaders = function (credentials, datetime) {
+  this.headers['X-Amz-Date'] = datetime;
```
Comment (eschultink) [inline S3Request.gs:221]: probably not relevant, but AWS example shows this in lowercase (`x-amz-date`), and I think it's worth being consistent at least amongst all `x-amz` headers.

File: S3Request.gs [hunk 2]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+      }
     }
   }
-  // (3) is just that multiple values of the same header must be passed as CSV, rather than listed multiple times; implicit
-  // sort lexographically (2), and combine into string w single \n separating each (6)
-  if (amzHeaders.length > 0) {
-    stringToSign += amzHeaders.sort().join("\n") + "\n";
+
+  var sep = this._getUrl().indexOf('?') >= 0 ? '&' : '?';
+  function queryParamsToString(params) {
```
Comment (eschultink) [inline S3Request.gs]: imho, better style to explicitly assign the function as a value of a local variable, rather than rely on people knowing JS scoping rules:

```suggestion
  var queryParamsToString = function(params) {
```

File: S3Request.gs [hunk 3]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
-      var header = headerName.toLowerCase() + ":" + this.headers[headerName].replace(/\s+/, " ");
-      amzHeaders.push(header) 
+S3Request.prototype.updateForPresigned = function(credentials, datetime) {
+  var credString = this.credentialString(datetime);
+  var qs = {
+    'X-Amz-Date': datetime,
+    'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
+    'X-Amz-Credential': credentials.accessKeyId + '/' + credString,
+    'X-Amz-Expires': this.headers[this.expiresHeader],
+    'X-Amz-SignedHeaders': this.signedHeaders()
```
Comment (eschultink) [inline S3Request.gs:234]: 🤔 mmm ... AWS seems to expect these as upper-case when creating canonical query string, but then lower case when actually put as headers on HTTP request.

File: S3Request.gs [hunk 4]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+    qs['Cache-Control'] = this.headers['Cache-Control'];
+  }
+
+  for (var key in this.headers) {
+    if (key === this.expiresHeader) return;
+    if (this.isSignableHeader(key)) {
+      var lowerKey = key.toLowerCase();
+      // Metadata should be normalized
+      if (lowerKey.indexOf('x-amz-meta-') === 0) {
+        qs[lowerKey] = value;
```
Comment (eschultink) [inline S3Request.gs]: I think `value` is undefined here. I think you need `var value = this.headers[key];` prior to this line or change this line to `qs[lowerKey] = this.headers[key];` (and line 252 below)

File: S3Request.gs [hunk 5]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+S3Request.prototype.isSignableHeader = function(key) {
+  if (key.match(/^x-amz/i)) return true;
+  var unsignableHeaders = [
+    'authorization',
+    'content-type',
+    'content-length',
+    'user-agent',
+    this.expiresHeader,
+    'expect',
+    'x-amzn-trace-id'
```
Comment (eschultink) [inline S3Request.gs:416]: this correct? seems inconsistent with how they name other headers

File: S3Request.gs [hunk 6]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+  ];
+  return unsignableHeaders.indexOf(key) < 0
```
Comment (eschultink) [inline S3Request.gs]: i think you need to lower-case `key` to test this correctly. HTTP headers are case insensitive. Also I think comparison would be better testing the whole string for equivalence, rather than whether header name begins with name of a header that you know to be unsignable.

File: S3Request.gs [hunk 7]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+
+  if (this.headers['Content-Type']) {
+    qs['Content-Type'] = this.headers['Content-Type'];
+  }
+  if (this.headers['Content-MD5']) {
+    qs['Content-MD5'] = this.headers['Content-MD5'];
+  }
+  if (this.headers['Cache-Control']) {
+    qs['Cache-Control'] = this.headers['Cache-Control'];
+  }
```
Comment (eschultink) [inline S3Request.gs:249]: seems weird to not have these cases covered by your loop below.  In essence, you're including all of these in the qs to be signed. so why not have `isSignableHeader()` return true for them? eg, you have `content-type` as something not to be signed in the body of `isSignableHeader()`

File: S3Request.gs [hunk 8]
```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
...
+
+  var sep = this._getUrl().indexOf('?') >= 0 ? '&' : '?';
+  function queryParamsToString(params) {
+    var items = [];
+    for (var key in params) {
+      var value = params[key];
+      var ename = encodeURIComponent(key);
+      if (Array.isArray(value)) {
+        var vals = [];
+        for(var val in value) {vals.push(encodeURIComponent(val))}
```
Comment (eschultink) [inline S3Request.gs]: this seems incorrect.  `val` here will be the index of the value in the array, not the value itself. so 
```suggestion
        for(var i in value) { vals.push(encodeURIComponent(value[i])) }
```

File: S3Request.gs [hunk 9]
```diff
@@ -95,35 +99,53 @@ S3Request.prototype.setObjectName = function(objectName) {
...
+  this.headers['Host'] = this._getUrl().replace(/https?:\/\/(.+amazonaws\.com).*/, '$1');
+
+  var credentials = {
+    accessKeyId: this.service.accessKeyId,
+    secretAccessKey: this.service.secretAccessKey,
+    sessionToken: options.sessionToken
+  }
+
+  this.addAuthorization(credentials, this.date)
+  // To avoid the error of confliction in UrlFetchApp#fetch. UrlFetchApp#fetch adds a Host header.
```
Comment (eschultink) [inline S3Request.gs]:
```suggestion
  // To avoid conflict with UrlFetchApp#fetch. UrlFetchApp#fetch adds a Host header.
```

**Review by eschultink — COMMENTED** [review]
thanks for your PR. Sorry I haven't had much time to maintain this lib, nor do I currently have a good method of testing out the functionality. I've tried to give a quick review and made a number of suggestions. In particular, I think there are a few bugs

File: S3Request.gs (hunk 1 above)
Comment (sakazuki) [reply S3Request.gs:221]: I can agree the consistent rule. But the master (AWS official) code [v4.js](https://github.com/aws/aws-sdk-js/blob/master/lib/signers/v4.js#L38-L43) uses this camelcase like rule.
We have to select which rule we use.

File: S3Request.gs (hunk 4 above)
Comment (sakazuki) [reply S3Request.gs]: You are right. I fix them

File: S3Request.gs (hunk 5 above)
Comment (sakazuki) [reply S3Request.gs:416]: I have not enough understanding about these part.  But I ported the master(AWS official) code [v4.js](https://github.com/aws/aws-sdk-js/blob/master/lib/signers/v4.js#L191-L199)

File: S3Request.gs (hunk 6 above)
Comment (sakazuki) [reply S3Request.gs]: I agree. I fix.

File: S3Request.gs (hunk 3 above)
Comment (sakazuki) [reply S3Request.gs:234]: I think this is the same as the first comment. Which rule we select ?

File: S3Request.gs (hunk 7 above)
Comment (sakazuki) [reply S3Request.gs:249]: I have not enough understanding about these part. But I ported the master(AWS official) code [v4.js](https://github.com/aws/aws-sdk-js/blob/master/lib/signers/v4.js#L65-L67) 
I think v4.js is a common library, and other service (not s3) requires these spec. (ex. Cloud Front processes the Cache-Control header and does not pass it to S3)

File: S3Request.gs (hunk 1 above)
Comment (eschultink) [reply S3Request.gs:221]: OK, I see. well then I have complaints about the quality of AMZN's code as well, which I guess isn't so surprising. I'm fine to leave it. should reference that you've ported that code to here in a comment somewhere to be clear.

File: S3Request.gs (hunk 7 above)
Comment (eschultink) [reply S3Request.gs:249]: yeah, I agree that AMZN's version of the lib has this structure problem too. It makes it hard to follow, and therefore hard to you to port over the apps script.

File: S3Request.gs [hunk 10]
```diff
@@ -242,27 +243,27 @@ S3Request.prototype.updateForPresigned = function(credentials, datetime) {
   }
 
   for (var key in this.headers) {
-    if (key === this.expiresHeader) return;
+    if (key === this.expiresHeader) continue;
```
Comment (eschultink) [inline S3Request.gs:252]: 👍

**Review by eschultink — APPROVED** [review]
generally looks OK, could be cleaner, but I agree that structure seems to follow AMZN version reasonably closely I think worth adding a comment referring to that lib, just to help people follow it more clearly - but otherwise OK!

File: S3Request.gs (hunk 1 above)
Comment (sakazuki) [reply S3Request.gs:221]: OK. I add reference comment in this code.

**Comment by sakazuki** [comment]
I add reference comment

File: S3Request.gs (hunk 7 above)
Comment (sakazuki) [reply S3Request.gs:249]: yeah,  I had tried to port the whole thing but gave up...

**Comment by sakoht** [comment]
This is great.  It looks like the one on https://script.google.com is the previous version.  Is there a way to reference this newer version?

@ sakazuki perhaps you have it up?
